File: nodejs/src/dataHandler.js
const fs = require('fs/promises');
const path = require('path');
const ExcelJS = require('exceljs');

class DataHandler {
    constructor(executionPath, logger) {
        this.executionPath = executionPath;
        this.initialDir = path.join(executionPath, 'dados_extraidos');
        this.processedDir = path.join(executionPath, 'dados_processados');
        this.logger = logger.child({ service: 'DataHandler' });
        this._createDirectories();
    }

    async _createDirectories() {
        try {
            await fs.mkdir(this.initialDir, { recursive: true });
            await fs.mkdir(this.processedDir, { recursive: true });
            this.logger.info("Diretórios de dados verificados/criados.");
        } catch (error) {
            this.logger.error(`Falha ao criar diretórios de dados: ${error.message}`);
            throw error;
        }
    }

    async saveSingleMovieAsJson(movieData, fileId) {
        if (!movieData || !movieData['nome_filme']) {
            this.logger.warn("Dados do filme inválidos ou sem nome, não será salvo.");
            return false;
        }

        const filename = `movie_${fileId}.json`;
        const filepath = path.join(this.initialDir, filename);

        try {
            const jsonString = JSON.stringify(movieData, null, 4);
            await fs.writeFile(filepath, jsonString, 'utf-8');
            this.logger.debug(`Dados do filme salvos em: ${filepath}`);
            return true;
        } catch (error) {
            this.logger.error(`Falha ao salvar o arquivo JSON para o filme ${movieData['nome_filme']}: ${error.message}`);
            return false;
        }
    }

    async consolidateJsonToExcel(outputFilename) {
        let fileList;
        try {
            fileList = await fs.readdir(this.initialDir);
        } catch (error) {
            this.logger.error(`Não foi possível ler o diretório 'dados_extraidos': ${error.message}`);
            return;
        }

        const jsonFiles = fileList.filter(f => f.endsWith('.json'));

        if (jsonFiles.length === 0) {
            this.logger.warn("Nenhum arquivo JSON encontrado para consolidar.");
            return;
        }

        const allMoviesData = [];
        for (const file of jsonFiles) {
            const filepath = path.join(this.initialDir, file);
            try {
                const content = await fs.readFile(filepath, 'utf-8');
                allMoviesData.push(JSON.parse(content));
            } catch (error) {
                this.logger.error(`Erro ao ler o arquivo JSON: ${filepath}. Pulando.`);
            }
        }

        if (allMoviesData.length === 0) {
            this.logger.error("Nenhum dado válido foi carregado dos arquivos JSON.");
            return;
        }

        this.logger.info(`Consolidando ${allMoviesData.length} arquivos de filme em uma planilha Excel.`);
        
        const workbook = new ExcelJS.Workbook();
        const worksheet = workbook.addWorksheet('Top Filmes IMDB');

        worksheet.columns = Object.keys(allMoviesData[0]).map(key => ({
            header: key,
            key: key,
            width: key === 'nome_filme' ? 40 : (key === 'sinopse' ? 60 : 15)
        }));

        worksheet.addRows(allMoviesData);

        const excelPath = path.join(this.processedDir, outputFilename);
        try {
            await workbook.xlsx.writeFile(excelPath);
            this.logger.info(`Planilha Excel final gerada com sucesso em: ${excelPath}`);
        } catch (error) {
            this.logger.error(`Falha ao gerar o arquivo Excel final: ${error.message}`);
        }
    }
}

module.exports = DataHandler;

File: python/src/data_handler.py
import pandas as pd
import os
import logging
import json
import glob

logger = logging.getLogger(__name__)

class DataHandler:
    def __init__(self, execution_path):
        self.execution_path = execution_path
        self.initial_dir = os.path.join(execution_path, 'dados_extraidos')
        self.processed_dir = os.path.join(execution_path, 'dados_processados')
        self._create_directories()

    def _create_directories(self):
        try:
            os.makedirs(self.initial_dir, exist_ok=True)
            os.makedirs(self.processed_dir, exist_ok=True)
            logger.info("Diretórios de dados verificados/criados.")
        except OSError as e:
            logger.error("Falha ao criar diretórios de dados: %s", e)
            raise

    def save_single_movie_as_json(self, movie_data, file_id):
        """Salva os dados de um único filme em um arquivo JSON."""
        if not movie_data or not movie_data.get('Nome do filme'):
            logger.warning("Dados do filme inválidos ou sem nome, não será salvo.")
            return False
        
        filename = f"movie_{file_id}.json"
        filepath = os.path.join(self.initial_dir, filename)

        try:
            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(movie_data, f, ensure_ascii=False, indent=4)
            logger.debug(f"Dados do filme salvos em: {filepath}")
            return True
        except Exception as e:
            logger.error(f"Falha ao salvar o arquivo JSON para o filme {movie_data.get('Nome do filme')}.", exc_info=True)
            return False

    def consolidate_json_to_excel(self, output_filename):
        """Lê todos os arquivos JSON, consolida e gera o Excel final."""
        json_pattern = os.path.join(self.initial_dir, '*.json')
        file_list = glob.glob(json_pattern)

        if not file_list:
            logger.warning("Nenhum arquivo JSON encontrado em 'dados_extraidos' para consolidar.")
            return

        all_movies_data = []
        for filepath in file_list:
            try:
                with open(filepath, 'r', encoding='utf-8') as f:
                    all_movies_data.append(json.load(f))
            except json.JSONDecodeError:
                logger.error(f"Erro ao ler o arquivo JSON: {filepath}. Pulando.")
        
        if not all_movies_data:
            logger.error("Nenhum dado válido foi carregado dos arquivos JSON.")
            return

        logger.info(f"Consolidando {len(all_movies_data)} arquivos de filme em uma planilha Excel.")
        df = pd.DataFrame(all_movies_data)
        
        excel_path = os.path.join(self.processed_dir, output_filename)
        try:
            df.to_excel(excel_path, index=False)
            logger.info(f"Planilha Excel final gerada com sucesso em: {excel_path}")
        except Exception as e:
            logger.error("Falha ao gerar o arquivo Excel final.", exc_info=True)